driver-core/src/main/java/com/datastax/driver/core/Message.java

```diff
@@ -192,21 +196,18 @@ public ProtocolEncoder(int version) {
             this.protocolVersion = version;
         }
 
-        @SuppressWarnings("unchecked")
-        public Object encode(ChannelHandlerContext ctx, Channel channel, Object msg) {
-            assert msg instanceof Request : "Expecting request, got " + msg;
-
-            Request request = (Request)msg;
-
+        @Override
+        protected void encode(ChannelHandlerContext ctx, Request request, List<Object> out) throws Exception {
             EnumSet<Frame.Header.Flag> flags = EnumSet.noneOf(Frame.Header.Flag.class);
             if (request.isTracingRequested())
                 flags.add(Frame.Header.Flag.TRACING);
 
+            @SuppressWarnings("unchecked")
             Coder<Request> coder = (Coder<Request>)request.type.coder(protocolVersion);
-            ChannelBuffer body = ChannelBuffers.buffer(coder.encodedSize(request));
+            ByteBuf body = ctx.alloc().ioBuffer(coder.encodedSize(request));
```

Comment:
I was thinking "it's something we're sending so that's I/O", but I overlooked the fact that we have more handlers downstream. Will change it.